Add migration guide from Ingress (#1547)
kubernetes-sigs/gateway-api: site-src/guides/migrating-from-ingress.md

# Migrating from Ingress

> The Gateway API is not backward compatible with the [Ingress API](https://kubernetes.io/docs/concepts/services-networking/ingress/). As a result, if you are an Ingress user and would like to migrate, you will need to convert your Ingress resources to Gateway API resources. 

Reviewer: Although this is largely true, it feels like a fairly negative way to start this guide. Maybe focus on the idea that Gateway API can generally be considered a superset of Ingress and is capable of much more than the original API.

Reviewer: LGTM, thanks @shaneutt!

Author: thanks! I included the suggestion with some edits.

This guide will help you with the conversion. It will:

> * Explain the key differences between the Ingress API and the Gateway API.

Reviewer: Maybe it's also worth adding a section for why you may want to switch to Gateway API.

> * Map the Ingress features to the Gateway API features.

Reviewer 2: tiny nit

```suggestion
* Map Ingress features to Gateway API features.
```

* Show an example of an Ingress resource converted to Gateway API resources.
* Mention a helper tool for automatic conversion.

> At the same time, it will not prepare you for a live migration or explain how to convert some vendor-specific features of your Ingress controller. Additionally, since the Ingress API only covers HTTP/HTTPS traffic, this guide does not cover the Gateway API support for other protocols.

Reviewer: Tiny nit
```suggestion
At the same time, it will not prepare you for a live migration or explain how to convert some implementation-specific features of your Ingress controller. Additionally, since the Ingress API only covers HTTP/HTTPS traffic, this guide does not cover the Gateway API support for other protocols.
```

## Key Differences Between Ingress API and Gateway API

There are three major differences between the Ingress API and the Gateway API:

* Personas.
* Available features.

> * Approach to extensibility (vendor-specific features).

Reviewer 2: ```suggestion
* Approach to extensibility (implementation-specific features).
```

### Personas

> At first, the Ingress API had only a single resource kind Ingress. As a result, it had only one persona -- the user -- the owner of Ingress resources. The Ingress features give the user a lot of control over how applications are exposed to their external clients, including TLS termination configuration and provisioning of the load balancing infrastructure (supported by some Ingress controllers). Such a level of control is called the self-service model. 

Reviewer: Tiny nits:

1. If you can wrap lines at 80 chars it makes it easier to leave comments/review smaller chunks at a time.
2. The first sentence feels just a bit off to me, maybe "Initially the Ingress API only had a single kind of resource."

* The cluster operator/application admin defines entry points for the external client traffic including TLS termination configuration.
* The application developer defines routing rules for their applications that  attach to those entry points.

Such a split adheres to a common organizational structure where multiple teams share the same load-balancing infrastructure. At the same time, it is not mandatory to give up the self-service model -- it is still possible to configure a single RBAC Role that will fulfill the application developer, application admin, and cluster operator responsibilities.

The table below summarizes the mapping between the Ingress API and the Gateway API personas:

| Ingress API Persona | Gateway API Persona |
|-|-|
| User | Application developer, Application admin, Cluster operator |
| Cluster operator | Cluster operator |
| Infrastructure provider | Infrastructure provider | 

### Available Features 

> The Ingress API comes with basic features only: TLS termination and content-based routing of HTTP traffic based on the host header and the URI of a request. To offer more features, Ingress controllers support them through annotations on the Ingress resource which are vendor-specific extensions to the Ingress API.

Reviewer 2: ```suggestion
The Ingress API comes with basic features only: TLS termination and content-based routing of HTTP traffic based on the host header and the URI of a request. To offer more features, Ingress controllers support them through annotations on the Ingress resource which are implementation-specific extensions to the Ingress API.
```

The Gateway API supports all the features of the Ingress resources and many features that are only available through annotations. As a result, the Gateway API is more portable than the Ingress API. Additionally, as the next section will show, you will not need to use any annotations at all, which addresses the awkwardness problem. 

### Approach to Extensibility

The Ingress API has two extensions points:

* Annotations on the Ingress resource (described in the previous section).
* [Resource backends](https://kubernetes.io/docs/concepts/services-networking/ingress/#resource-backend), which is the ability to specify a backend other than a Service.

The Gateway API is feature-rich compared with the Ingress API. However, to configure some advanced features like authentication or common but non-portable across data planes features like connection timeouts and health checks, you will need to rely on the extensions of the Gateway API.

The Gateway API has the following primary extension points:

> * *External references.* A feature (field) of a Gateway API resource can reference a custom resource (CRD) specific to the Gateway implementation that configures that feature. For example:

Reviewer: Tiny nit, I'd either remove "CRD" or replace it with "CR" to more accurately represent the "custom resource" terminology:

```suggestion
* *External references.* A feature (field) of a Gateway API resource can reference a custom resource specific to the Gateway implementation that configures that feature. For example:
```

    * [BackendObjectReference](/references/spec/#gateway.networking.k8s.io/v1beta1.BackendObjectReference) supports resources other than Services.
    * [SecretObjectReference](/references/spec/#gateway.networking.k8s.io/v1beta1.SecretObjectReference) supports resources other than Secrets.
* *Custom implementations*. For some features, it is left up to an implementation to define how to support them. Those features correspond to the implementation-specific (custom)  [conformance level](/concepts/conformance/#2-support-levels). For example:
    * The `RegularExpression` type of the [HTTPPathMatch](/references/spec/#gateway.networking.k8s.io/v1beta1.SecretObjectReference).
* *Policies.* A Gateway implementation can define custom resources (CRDs) called Policies for exposing data plane features like authentication. The Gateway API does not prescribe the details of those resources. However, it prescribes a standard UX. See the [Policy attachment guide](/references/policy-attachment/) for more details. In contrast with the *external references* above, a Gateway API resource does not reference a Policy. Instead, a Policy must reference a Gateway API resource.

> The extension points do not include annotations on the Gateway API resources. This approach is strongly discouraged for Gateway implementations.

Reviewer: ```suggestion
The extension points do not include annotations on the Gateway API resources. This approach is strongly discouraged for implementations of the API.
```

## Mapping Ingress API features to Gateway API Features

This section will map the Ingress API features to the corresponding Gateway API features. It will cover three major areas:

* Entry points.
* TLS termination.

> * Routing rules.

Reviewer: Tiny nit: I don't think these need to end with `.` because they're not sentences.

### Entry Points

Roughly speaking, an entry point is a combination of an IP address and port through which the data plane is accessible to external clients.

Every Ingress resource has two implicit entry points -- one for HTTP and the other for HTTPS traffic. An Ingress controller provides those entry points. Typically, they will either be shared by all Ingress resources, or every Ingress resource will get dedicated entry points.

In the Gateway API, entry points must be explicitly defined in a [Gateway](/api-types/gateway/) resource. For example, if you want the data plane to handle HTTP traffic on port 80, you need to define a [listener](/references/spec/#gateway.networking.k8s.io/v1beta1.Listener) for that traffic. Typically, a Gateway implementation provides a dedicated data plane for each Gateway resource.

Gateway resources are owned by the cluster operator and the application admin.

### TLS Termination

The Ingress resource supports TLS termination via the [TLS section](https://kubernetes.io/docs/concepts/services-networking/ingress/#tls), where the TLS certificate and key are stored in a Secret.

> In the Gateway API, TLS termination is a property of the [listener](/references/spec/#gateway.networking.k8s.io/v1beta1.Listener), and similarly to the Ingress, a TLS certificate and key are also stored in a Secret.

Reviewer: ```suggestion
In the Gateway API, TLS termination is a property of the [Gateway listener](/references/spec/#gateway.networking.k8s.io/v1beta1.Listener), and similarly to the Ingress, a TLS certificate and key are also stored in a Secret.
```

Because the listener is part of the Gateway resource, the cluster operator and application admin own TLS termination.

### Routing Rules

The [path-based routing rules](https://kubernetes.io/docs/concepts/services-networking/ingress/#path-types) of the Ingress resource map directly to the [routing rules](/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRouteRule) of the [HTTPRoute](/api-types/httproute/).

> The [host-header-based routing rules](https://kubernetes.io/docs/concepts/services-networking/ingress/#name-based-virtual-hosting) map to the [hostnames](/references/spec/#gateway.networking.k8s.io/v1beta1.Hostname) of the HTTPRoute. However, note that in the Ingress, each hostname has separate routing rules, while in the HTTPRoute the routing rules apply to all hostnames.

Reviewer 2: Hostnames can also be set on Gateway Listeners which should probably be mentioned here as well

Typically, Ingress controllers merge routing rules from all Ingress resources (unless they provision a data plane per each Ingress resource) and resolve potential conflicts among the rules. However, both merging and conflict resolution are not prescribed by the Ingress API, so Ingress controllers might implement them differently.

In contrast, the Gateway API specifies how to merge rules and resolve conflicts:

* A Gateway implementation must merge the routing rules from all HTTPRoutes attached to a listener.
* Conflicts must be handled as prescribed [here](/concepts/guidelines/#conflicts). For example, more specific matches in a routing rule win over the less specific ones.

#### Default Backend

The Ingress [default backend](https://kubernetes.io/docs/concepts/services-networking/ingress/#default-backend) configures a backend that will respond to all unmatched HTTP requests related to that Ingress resource. The Gateway API does not have a direct equivalent: it is necessary to define such a routing rule explicitly. For example, define a rule to route requests with the path prefix `/` to a Service that corresponds to the default backend.

#### Selecting Data Plane to Attach to

> An Ingress resource must specify a [class](https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-class) to select which Ingress controller to use. An HTTPRoute must specify which Gateway (or multiple Gateways) to attach to via a [parentRef](/references/spec/#gateway.networking.k8s.io/v1beta1.ParentRef).

Reviewer: ```suggestion
An Ingress resource must specify a [class](https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-class) to select which Ingress controller to use. An HTTPRoute must specify which Gateway (or Gateways) to attach to via a [parentRef](/references/spec/#gateway.networking.k8s.io/v1beta1.ParentRef).
```

### Vendor-Specific Ingress Features (Annotations) 

Ingress annotations configure vendor-specific features. Thus, converting them to the Gateway API depends both on the Ingress controller and Gateway implementation vendors.

Luckily, some of the features supported through annotations are now part of the Gateway API (HTTPRoute), primarily:

* Request redirects (including a TLS redirect).
* Request/response manipulation.
* Traffic splitting.

> * Header, cookie, or method-based routing.

Reviewer 2: I don't think cookie based routing is available in GW API (yet)

    secretName: example-com
  rules:
  - host: foo.example.com
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: foo-app
            port:
              number: 80
      - path: /orders
        pathType: Prefix 
        backend:
          service:
            name: foo-orders-app
            port:
              number: 80
  - host: bar.example.com
    http:
      paths:
      - path: /
        pathType: Prefix 
        backend:
          service:
            name: bar-app
            port:
              number: 80
```

The next three sections convert that Ingress into Gateway API resources.

### Conversion Step 1 - Define Gateway 

The following Gateway resource:

* Belongs to our GatewayClass `prod`.
* Provisions load balancing infrastructure (this depends on the Gateway implementation).
* Configures HTTP and HTTPS listeners (entry points), which the Ingress resource included implicitly:
    * The HTTP listener `http` on port `80`. 
    * The HTTPS listener `https` on port `443` with TLS termination with the cert and key stored in the `example-com` Secret. It is the same Secret used in the Ingress. 

> Also, note that both listeners allow all HTTPRoutes from the same namespace (which is the default setting) and restrict HTTPRoute hostnames to the `example.com` subdomain.

Reviewer 2: if we include this detail and the `*.example.com` configuration we should describe how this and the hostname configuration on HTTPRoutes works (intersection etc.)

# Migrating from Ingress

> The Gateway API project is the successor to the Ingress API. However, it does
> not include the Ingress resource (the closest parallel is the HTTPRoute). As a
> result, a one-time conversion from your existing Ingress resources to the
> relevant Gateway API resources is necessary.

Reviewer 3: We link to `Ingress` below, but since we're basically "introducing it" here we should have a link up here as well. Also a language nit:

```suggestion
The Gateway API project is the successor to the [Ingress][ing] API. However, it
does not include the `Ingress` resource (the closest parallel is the `HTTPRoute`
resource). As a result, a one-time conversion from your existing Ingress resources
to the relevant Gateway API resources is necessary.

[ing]:https://kubernetes.io/docs/concepts/services-networking/ingress/
```

  for automatic conversion.

At the same time, it will not prepare you for a live migration or explain how to
convert some implementation-specific features of your Ingress controller.
Additionally, since the Ingress API only covers HTTP/HTTPS traffic, this guide
does not cover the Gateway API support for other protocols.

## Reasons to Switch to Gateway API

The [Ingress API](https://kubernetes.io/docs/concepts/services-networking/ingress/)
is the standard Kubernetes way to configure external HTTP/HTTPS load balancing
for Services. It is widely adopted by Kubernetes users and well-supported by
vendors with many implementations ([Ingress controllers](https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/))
available. Additionally, several cloud-native projects integrate with the
Ingress API, such as [cert-manager](https://cert-manager.io/)
and [ExternalDNS](https://github.com/kubernetes-sigs/external-dns).

However, the Ingress API has several limitations:

- *Limited features*. The Ingress API only supports TLS termination and
  simple content-based request routing of HTTP traffic.

> - *Reliance on annotations for extensibility*. The annotations approach to
>   extensibility leads to the limited portability and awkwardness of the API. 

Reviewer 3: I feel like "awkwardness" here leaves a lot to decipher for the reader. Perhaps we can dig in more:

```suggestion
- *Reliance on annotations for extensibility*. The annotations approach to
  extensibility leads to limited portability as every implementation has its own
  supported extensions that may not translate to any other implementation.
```

* The cluster operator/application admin defines entry points for the external
  client traffic including TLS termination configuration.
* The application developer defines routing rules for their applications that
  attach to those entry points.

Such a split adheres to a common organizational structure where multiple teams
share the same load-balancing infrastructure. At the same time, it is not
mandatory to give up the self-service model -- it is still possible to configure
a single RBAC Role that will fulfill the application developer, application
admin, and cluster operator responsibilities.

The table below summarizes the mapping between the Ingress API and the Gateway
API personas:

| Ingress API Persona | Gateway API Persona |
|-|-|
| User | Application developer, Application admin, Cluster operator |
| Cluster operator | Cluster operator |
| Infrastructure provider | Infrastructure provider |

### Available Features

The Ingress API comes with basic features only: TLS termination and
content-based routing of HTTP traffic based on the host header and the URI of a
request. To offer more features, Ingress controllers support them through

> annotations on the Ingress resource which are implementation-specific extensions
> to the Ingress API.
> 

Reviewer 3: ```suggestion
[annotations][anns] on the Ingress resource which are implementation-specific
extensions to the Ingress API.

[anns]:https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/
```

      supports resources other than Secrets.
* *Custom implementations*. For some features, it is left up to an
  implementation to define how to support them. Those features correspond to the
  implementation-specific
  (custom)  [conformance level](/concepts/conformance/#2-support-levels). For
  example:
    * The `RegularExpression` type of
      the [HTTPPathMatch](/references/spec/#gateway.networking.k8s.io/v1beta1.SecretObjectReference).
* *Policies.* A Gateway implementation can define custom resources called
  Policies for exposing data plane features like authentication. The Gateway API
  does not prescribe the details of those resources. However, it prescribes a
  standard UX. See the [Policy attachment guide](/references/policy-attachment/)
  for more details. In contrast with the *external references* above, a Gateway
  API resource does not reference a Policy. Instead, a Policy must reference a
  Gateway API resource.

The extension points do not include annotations on the Gateway API resources.
This approach is strongly discouraged for implementations of the API.

## Mapping Ingress API features to Gateway API Features

> This section will map the Ingress API features to the corresponding Gateway API
> features. It will cover three major areas:

Reviewer 3: ```suggestion
This section will map the Ingress API features to the corresponding Gateway API
features, covering three major areas:
```

not `foo.kubernetes.io`).

```yaml
{% include 'standard/simple-http-https/gateway.yaml' %}
```

### Conversion Step 2 - Define HTTPRoutes

The Ingress is split into two HTTPRoutes -- one for `foo.example.com` and one
for `bar.example.com` hostnames.

```yaml
{% include 'standard/simple-http-https/foo-route.yaml' %}
```

```yaml
{% include 'standard/simple-http-https/bar-route.yaml' %}
```

Both HTTPRoutes:

* Attach to the `https` listener of the Gateway resource from Step 1.
* Define the same routing rules as in the Ingress rules for the corresponding
  hostname.

### Step 3 - Configure TLS Redirect

The following HTTPRoute configures a TLS redirect, which the Ingress resource
configured via an annotation. The HTTPRoute below:

* Attaches to the `http` listener of our Gateway.
* Issues a TLS redirect for any HTTP request for the `foo.example.com`
  or `bar.example.com` hostnames.

```yaml
{% include 'standard/simple-http-https/tls-redirect-route.yaml' %}
```

## Automatic Conversion of Ingresses

The [Ingress to Gateway](https://github.com/kubernetes-sigs/ingress2gateway)
project helps translate Ingress resources to Gateway API resources, specifically

> HTTPRoutes. It is early in the development phase and still experimental in
> nature. The conversion results should always be tested and verified.

Reviewer 3: I think I may have suggested this language, but in retrospect I feel like this comment will bitrot, I think it would be reasonable to drop and it and just let them follow the link to see the documentation of `ingress2gateway` which can inform them there of the current maturity and support levels:

```suggestion
HTTPRoutes. The conversion results should always be tested and verified.
```
